File: server/common/protocol/agency_handler.py
import logging

from common.network.connection_interface import ConnectionInterface
from common.protocol.bet_parser import BetParser
from common.utils.utils import Bet
from common.protocol.protocol_constants import *


class AgencyHandler:
    """Handles individual client connections"""

    def __init__(self):
        self.bet_parser = BetParser()

    def get_bets(self, connection: ConnectionInterface) -> list[Bet]:

        self._process_header(connection)

        bet = self.bet_parser.parse_bet(connection)
        return bet

    def _process_header(self, client_connection: ConnectionInterface):

        headerData = client_connection.receive(HEADER_SIZE)

        header_byte = headerData[0:1]

        if header_byte != BET_HEADER:
            raise Exception(f"Unexpected header: {header_byte}, expected: {BET_HEADER}")

        logging.debug(f"action: process_header | result: success")

    def confirm_bet(
        self, bets: list[Bet], connection: ConnectionInterface, status: bool
    ) -> None:
        """
        Confirm the bet with the client
        """
        try:
            if status:
                connection.send(SUCCESS_HEADER)

                bet = bets[0]
                bet_message = f"{bet.document},{bet.number},"
                message_bytes = bet_message.encode()
                message_length = len(message_bytes)
                length_bytes = bytes([message_length])

                connection.send(length_bytes)
                connection.send(message_bytes)

                logging.debug(f"action: confirm_bet | result: success")
            else:
                connection.send(FAIL)
                logging.debug(f"action: confirm_bet | result: fail")
        except Exception as e:
            logging.error(f"action: confirm_bet | result: fail | error: {e}")

File: client/common/protocol/agency_handler.go
package protocol

import (
	"encoding/binary"
	"fmt"

	"github.com/7574-sistemas-distribuidos/docker-compose-init/client/common/network"
	"github.com/op/go-logging"
)

var log = logging.MustGetLogger("log")

type AgencyHandler struct {
	MaxBatchAmount int
}

func NewAgencyHandler() *AgencyHandler {
	return &AgencyHandler{}
}

func (a *AgencyHandler) SendBets(bets string, connSock *network.ConnectionInterface) error {

	if len(bets) > MAX_BATCH_SIZE {
		return fmt.Errorf("bets size too big to send: %d bytes (max %d)", len(bets), MAX_BATCH_SIZE)
	}

	err := connSock.SendData([]byte(HEADER))
	if err != nil {
		return err
	}

	lenBytes := make([]byte, 2)
	binary.BigEndian.PutUint16(lenBytes, uint16(len(bets)))
	err = connSock.SendData(lenBytes)
	if err != nil {
		return err
	}
	err = connSock.SendData([]byte(bets))
	if err != nil {
		return err
	}
	log.Debug("action: send_bet_batch | result: success | batch_length: %d Bytes", len(bets))

	return err
}

func (a *AgencyHandler) RecvConfirmation(connSock *network.ConnectionInterface) error {
	headerData := make([]byte, SUCCESS_HEADER_SIZE)
	err := connSock.ReceiveData(headerData)
	if err != nil {
		return fmt.Errorf("failed to receive header: %v", err)
	}
	success_header := string(headerData)

	if success_header == SUCCESS_HEADER {
		log.Debug("action: batch_confirmation | result: success")
	} else {
		log.Errorf("action: batch_confirmation | result: fail")
		return fmt.Errorf("batch processing failed")
	}

	return nil
}

func (a *AgencyHandler) SendDone(connSock *network.ConnectionInterface) error {
	err := connSock.SendData([]byte(EOF))
	return err
}

func (a *AgencyHandler) GetResults(connSock *network.ConnectionInterface) (string, error) {
	log.Debug("action: waiting_lottery_results | result: success")
	headerData := make([]byte, WINNER_HEADER_SIZE)
	err := connSock.ReceiveData(headerData)
	if err != nil {
		return "", fmt.Errorf("failed to receive header: %v", err)
	}
	success_header := string(headerData)

	log.Debug("action: lottery_results_received | result: success")

	if success_header == WINNERS_HEADER {
		winners_data_count := make([]byte, WINNER_COUNT_SIZE)
		err := connSock.ReceiveData(winners_data_count)
		if err != nil {
			return "", fmt.Errorf("failed to receive length data: %v", err)
		}
		winnerCount := binary.BigEndian.Uint16(winners_data_count)

		winnerData := make([]byte, winnerCount)
		err = connSock.ReceiveData(winnerData)
		if err != nil {
			return "", fmt.Errorf("failed to receive winner data: %v", err)
		}
		winnerStr := string(winnerData)

		return winnerStr, nil

	} else {
		log.Errorf("action: winner_confirmation | result: fail")
		return "", fmt.Errorf("an error occurred tallying up the winners")
	}
}
